httpx/concurrency/asyncio.py

```diff
@@ -194,6 +194,44 @@ def loop(self) -> asyncio.AbstractEventLoop:
             stream_reader=stream_reader, stream_writer=stream_writer, timeout=timeout
         )
 
+    async def start_tls(
+        self,
+        stream: BaseStream,
+        hostname: str,
+        ssl_context: ssl.SSLContext,
+        timeout: TimeoutConfig,
+    ) -> BaseStream:
+
+        loop = self.loop
+        if not hasattr(loop, "start_tls"):  # pragma: no cover
+            raise NotImplementedError(
+                "asyncio.AbstractEventLoop.start_tls() is only available in Python 3.7+"
```

Comment:
Are we fine not supporting this on 3.6?

I'm not familiar with the workings of TLS but looking around in aiohttp it seems [they use `loop.create_connection`](https://github.com/aio-libs/aiohttp/blob/8742305de4ca7b159dc452697802e9a05309f298/aiohttp/connector.py#L1061). Since we're creating the stream reader and writer here maybe we can do something similar with [`asyncio.open_connection`](https://docs.python.org/3/library/asyncio-stream.html#asyncio.open_connection) and avoid using `start_tls`?

Sorry if I'm missing something obvious, like I said I'm no expert in TLS 🙂

Reply:
From what I understand, start_tls allows to « upgrade » an existing connection to be encrypted when it was not previously — notice how we reuse the stream’s transport in the code here.

There may be a workaround, but at first glance I would guess that basing that off open_connection would open another socket, which is not what we want (we want to keep the existing connection and upgrade it over TLS).

From a Python compatibility point of view, the main implication is that HTTP Tunnel Proxy would only be available on 3.7, correct @sethmlarson? (Then I’m thinking maybe it will be worth catching the exception raised here to provide a more user-friendly message.)

Reply:
@florimondmanca is correct, this method takes any steam and performs the TLS handshake, even if it already has TLS active!

And you're right that we won't be able to support the HTTP tunnel on Python 3.6 with the asyncio backend. The warning message can recommend either upgrading Python or using the Trio backend perhaps.